opm/grid/cpgrid/processEclipseFormat.cpp

```diff
@@ -638,39 +670,126 @@ namespace cpgrid
-            if (nnc.size() > 0) {
-                int cart_size = 1;
-                int num_dims = sizeof(output.dimensions)/ sizeof(*output.dimensions);
-                for (int idx = 0; idx < num_dims ; ++idx) {
-                    cart_size *= output.dimensions[idx];
+            int cart_size = 1;
+            const int num_dims = sizeof(output.dimensions)/sizeof(*output.dimensions);
+            for (int idx = 0; idx < num_dims ; ++idx) {
+                cart_size *= output.dimensions[idx];
+            }
+            // create the inverse map of global_cell
+            global_to_local.resize(cart_size, -1);
+            for (size_t idx = 0; idx < global_cell.size(); ++idx) {
+                global_to_local[global_cell[idx]] = idx;
+            }
+            return global_to_local;
+        }
+
+
+
+
+
+        NNCMap filterNNCs(const processed_grid& output,
+                          const NNCMap& nnc,
+                          const std::vector<int>& global_to_local)
+        {
+            NNCMap filtered_nnc;
+            const int num_faces = output.number_of_faces;
+            std::vector<std::pair<int, int>> face_cells(num_faces);
+            // Sort all face->cell mappings so that lowest cell number comes first.
+            for (int f = 0; f < num_faces; ++f) {
+                const int c1 = output.face_neighbors[2*f];
+                const int c2 = output.face_neighbors[2*f + 1];
+                if (c1 < c2) {
+                    face_cells[f] = { c1, c2 };
+                } else {
+                    face_cells[f] = { c2, c1 };
                 }
-                // create the inverse map of global_cell
-                global_to_local.resize(cart_size, -1);
-                for (size_t idx = 0; idx < global_cell.size(); ++idx) {
-                    global_to_local[global_cell[idx]] = idx;
+            }
+            // Sort face->cell mappings according to first, then second cell.
+            std::sort(face_cells.begin(), face_cells.end());
+            // For each nnc, add it to filtered_nnc only if not found in face->cell mappings.
+            for (const auto& nncpair : nnc) {
+                if (nncpair.first >= global_to_local.size() || nncpair.second >= global_to_local.size()) {
+                    Opm::OpmLog::warning("nnc_invalid", "NNC connection requested between invalid cells.");
+                    continue;
+                }
+                const int c1 = global_to_local[nncpair.first];
+                const int c2 = global_to_local[nncpair.second];
+                if (c1 < 0 || c2 < 0) {
+                    Opm::OpmLog::warning("nnc_inactive", "NNC connection requested between inactive cells.");
+                    continue;
+                }
+                auto beg = std::lower_bound(face_cells.begin(), face_cells.end(), std::make_pair(c1, -1));
+                auto end = std::upper_bound(face_cells.begin(), face_cells.end(), std::make_pair(c1 + 1, -1));
```

Comment:
Do we really want to binary searches over all face_cells?

There should not be two many faces for a cell and therefore the second one could be linear:
```
auto end=std::find(begin, face_cells.end(), [](const auto& pair){return pair->first==c1+1;})
```
As `nnc` is sorted the first one could start with the last end.